clusterloader2/pkg/measurement/common/simple/service_creation_latency.go

```diff
@@ -0,0 +1,538 @@
+		case <-time.After(defaultWaitTimeout):
+			klog.Errorf("%s: waiting for services timeout")
+			return fmt.Errorf("waiting for services timeout")
+		case <-time.After(defaultCheckInterval):
+			func() {
+				s.lock.Lock()
+				defer s.lock.Unlock()
+				for _, ct := range s.creationTimes {
+					created++
+					if !ct.assigned.IsZero() {
+						assigned++
+					}
+					if !ct.available.IsZero() {
+						available++
+					}
+				}
+			}()
+			klog.Infof("%s: %d created, %d assigned, %d available", s, created, assigned, available)
+			if created == available {
+				return nil
+			}
+		}
+	}
+}
+
+func (s *serviceCreationLatencyMeasurement) gather() ([]measurement.Summary, error) {
+	klog.Infof("%s: gathering service creation latency measurement...", s)
+	if !s.isRunning {
+		return nil, fmt.Errorf("metric %s has not been started", s)
+	}
+
+	createToAssignedLatency := make([]serviceCreationLatencyData, 0)
+	assignedToAvailableLatency := make([]serviceCreationLatencyData, 0)
+	createToAvailableLatency := make([]serviceCreationLatencyData, 0)
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	for key, ct := range s.creationTimes {
+		if ct.creation.IsZero() || ct.assigned.IsZero() {
+			continue
+		}
+		createToAssignedLatency = append(createToAssignedLatency, serviceCreationLatencyData{
+			Key:     key,
+			Latency: ct.assigned.Sub(ct.creation),
+		})
+		if ct.available.IsZero() {
+			continue
+		}
+		assignedToAvailableLatency = append(assignedToAvailableLatency, serviceCreationLatencyData{
+			Key:     key,
+			Latency: ct.available.Sub(ct.assigned),
+		})
+		createToAvailableLatency = append(createToAvailableLatency, serviceCreationLatencyData{
+			Key:     key,
+			Latency: ct.available.Sub(ct.creation),
+		})
+	}
+
+	sort.Sort(serviceCreationLatencySlice(createToAssignedLatency))
+	sort.Sort(serviceCreationLatencySlice(assignedToAvailableLatency))
+	sort.Sort(serviceCreationLatencySlice(createToAvailableLatency))
+
+	latency := &serviceCreationLatency{
+		CreateToAssignedLatency:    extractLatencyMetrics(createToAssignedLatency),
+		AssignedToAvailableLatency: extractLatencyMetrics(assignedToAvailableLatency),
+		CreateToAvailableLatency:   extractLatencyMetrics(createToAvailableLatency),
+	}
+	content, err := util.PrettyPrintJSON(podStartupLatencyToPerfData(latency))
+	if err != nil {
+		return nil, err
+	}
+	summary := measurement.CreateSummary(serviceCreationLatencyName, "json", content)
+	return []measurement.Summary{summary}, err
+}
+
+type serviceCreationLatencyData struct {
+	Key     string
+	Latency time.Duration
+}
+
+type serviceCreationLatencySlice []serviceCreationLatencyData
+
+func (a serviceCreationLatencySlice) Len() int           { return len(a) }
+func (a serviceCreationLatencySlice) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a serviceCreationLatencySlice) Less(i, j int) bool { return a[i].Latency < a[j].Latency }
+
+func extractLatencyMetrics(latencies []serviceCreationLatencyData) measurementutil.LatencyMetric {
```

Comment:
The exact same function already exists in pod_startup_slo - it should be extracted somewhere.

Reply:
Done.